### PR TITLE
Docs: Update monitoring documentation for grafana and prometheus

### PR DESCRIPTION
**Pull request by h314to**

### CODE REVIEW

File: docs/configuration/monitoring.md [hunk 1]
```diff
@@ -12,32 +12,51 @@ we have all functionality available on Grafana.
 
 The simplest way to setup prometheus in your cluster is by using the
 [Prometheus-Operator](https://github.com/helm/charts/tree/master/stable/prometheus-operator)
-bundle.
+bundle. Start by adding the custom resources required for installing this bundle in your cluster.
 
 ```sh
+# NOTE: if installing on microk8s use `microk8s.kubectl` instead of `kubectl`
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: I would place the *NOTE:* outside the code block because it will be more visible. Later I will try to find a way to format all note / warning / ... text so it is more user-friendly and looks nicer.  :)

File: docs/configuration/monitoring.md [hunk 2]
```diff
@@ -12,32 +12,51 @@ we have all functionality available on Grafana.
...
 
-helm upgrade --install monitoring stable/prometheus-operator \
-  --version 6.9.3 \
-  --namespace monitoring \
-  --set prometheusOperator.createCustomResource=false \
-  --set grafana.service.type=NodePort\
-  --set grafana.adminPassword="CHANGE_HERE"
+You can now install the monitoring bundle on your cluster:
+
+```sh
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: `sh` works just as fine as `bash`, but for the sake of consistency I suggest that we keep using "bash" throughout like in all other Markdown files in this repository.

File: docs/configuration/monitoring.md [hunk 3]
```diff
@@ -12,32 +12,51 @@ we have all functionality available on Grafana.
...
+helm upgrade --install --atomic promtail loki/promtail --version 0.13.0 \
+  --recreate-pods --namespace monitoring --values promtail-values.yaml
+
+helm upgrade --install --atomic monitoring stable/prometheus-operator --version 6.9.3 \
+  --recreate-pods --namespace monitoring --values prometheus-operator-values.yaml
 ```
 
-Now that you have prometheus and grafana you can enable `ServiceMonitors` and `GrafanaDashboards`
-for Codacy components:
+Now that you have prometheus and grafana installed you can enable `serviceMonitors`
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: Just a nitpick, but check the capitalization of the 3rd party tools whenever is makes sense (when we're not referring to a lowercase command or setting, for example):

```suggestion
Now that you have Prometheus and Grafana installed you can enable `serviceMonitors`
```

File: docs/configuration/git-providers/monitoring-conf/loki-values.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+resources:
```
Comment (pauloribeiro-codacy) [inline docs/configuration/git-providers/monitoring-conf/loki-values.yaml]: To consider: we are providing all the template .yaml files under the directory /codacy. Would it make sense to include these there as well, as it makes it easier for user to find all templates in the same place? I'm especially worried that right now these files are buried quite deep inside the docs directory. :)

Comment (paulopontesm) [inline docs/configuration/git-providers/monitoring-conf/loki-values.yaml]: The path for this file seems quite strange ` docs/configuration/git-providers/monitoring-conf/loki-values.yaml`

Why is this under "git-providers" ?

Comment (paulopontesm) [reply docs/configuration/git-providers/monitoring-conf/loki-values.yaml]: What we are doing in the documentation for this is installing these components as a separate thing from codacy.

This line installs loki by itself and in a different namespace different than codacy:
`helm upgrade --install --atomic loki loki/loki --version 0.17.0 \
  --recreate-pods --namespace monitoring --values loki-values.yaml`

This "yaml" file is used to configure this new system.

File: docs/configuration/git-providers/monitoring-conf/loki-values.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  limits_config:
+    cardinality_limit: 200000
+    ingestion_rate_mb: 15
+    ingestion_burst_size_mb: 20
+serviceMonitor:
+  enabled: true
+persistence:
+  enabled: true
+  size: 100Gi # adjust persistence volume size if needed
+  storageClassName: <--- storage class ---> # use, for instance microk8s-hostpath, for microk8s
```
Comment (paulopontesm) [inline docs/configuration/git-providers/monitoring-conf/loki-values.yaml]: Wouldn't "default" be good enought here?

Ideally people wouldn't have to edit this field.

File: docs/configuration/monitoring.md [hunk 6]
```diff
@@ -52,6 +71,18 @@ engine:
...
 
+Apply this configuration by issuing an helm upgrade while passing these additional values.
+To do so append `--values monitoring-values.yaml --recreate-pods` to the command used to install
+codacy provided in [this documentation](../install.md).
+
+```sh
+helm upgrade (...options used to install codacy...) --values monitoring-values.yaml --recreate-pods
+```
+
+Grafana will be running in the same domain as Codacy, on port 30080, for instance
```
Comment (paulopontesm) [inline docs/configuration/monitoring.md]: How does this happen? We are not configuring the ingress for this, right?

File: docs/configuration/monitoring.md [hunk 7]
```diff
@@ -12,32 +12,51 @@ we have all functionality available on Grafana.
 
 The simplest way to setup prometheus in your cluster is by using the
 [Prometheus-Operator](https://github.com/helm/charts/tree/master/stable/prometheus-operator)
-bundle.
+bundle. Start by adding the custom resources required for installing this bundle in your cluster.
```
Comment (paulopontesm) [inline docs/configuration/monitoring.md]: I think we need to separate documents for the "crow" way of monitoring from the Prometheus + Loki + Grafana.

Also.. I believe it would be more clear if we just went through each of the components one by one:

# 1. Install Prometheus

> One liner saying what is prometheus

> A few line on how to install

# 2. Install Loki

# 3. Install Grafana

> Including how to configure the ingress to access it....

# 4. Enable Codacy ServiceMonitors and Dashboards

File: docs/configuration/monitoring.md [hunk 8]
```diff
@@ -75,20 +106,20 @@ Please see the [README.md](https://github.com/codacy/chart/blob/master/README.md
 
 You **should** provide a password for the `crow` installation either through the `values.yaml` file or through a `--set` parameter during the `helm` installation process. This parameter can be configured as follows:
 
-* Through a `--set` parameter:
+-   Through a `--set` parameter:
```
Comment (paulopontesm) [inline docs/configuration/monitoring.md]: You need to rebase.. there is a new config for remark that forces `*` instead of `-`

File: docs/configuration/git-providers/monitoring-conf/loki-values.yaml [hunk 9]
```diff
@@ -0,0 +1,24 @@
+resources:
+  limits:
```
Comment (paulopontesm) [inline docs/configuration/git-providers/monitoring-conf/loki-values.yaml]: Make it explicit in the docs how it affects resource usage

File: docs/configuration/git-providers/monitoring-conf/prometheus-operator-values.yaml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+      version: 1
+      orgId: 1
+      access: proxy
+      basicAuth: false
+      editable: true
+      jsonData:
+         tlsSkipVerify: true
+    - name: Prometheus
+      type: prometheus
+      url: http://prometheus-prometheus-oper-prometheus:9090/
```
Comment (paulopontesm) [inline docs/configuration/git-providers/monitoring-conf/prometheus-operator-values.yaml]: Doesn't this depend on the namespace?

File: docs/configuration/git-providers/monitoring-conf/loki-values.yaml (hunk 4 above)
Comment (h314to) [reply docs/configuration/git-providers/monitoring-conf/loki-values.yaml]: Yes, it is a good idea, but there's one caveat: the things inside `codacy/` get packaged into the chart. Not sure if it is wise to keep adding everything there. As the amount of extra stuff will keep growing that folder will get messier. For now it is probably fine, but in the future we should find a way to make it a bit more tidy. Maybe have a folder in the root of the repo where we collect all the extra stuff that most people won't need.

Comment (h314to) [reply docs/configuration/git-providers/monitoring-conf/loki-values.yaml]: Yes, it was wrong. Thanks. I'm moving it elsewhere, in keeping with @pauloribeiro-codacy 's suggestion.

File: docs/configuration/git-providers/monitoring-conf/loki-values.yaml (hunk 5 above)
Comment (h314to) [reply docs/configuration/git-providers/monitoring-conf/loki-values.yaml]: yes, that would work

File: docs/configuration/monitoring.md (hunk 6 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: Yeah, it was a nodeport, but that's messy. I'll remove it for now, and then use codacy ingress when we add it to the chart.

File: docs/configuration/git-providers/monitoring-conf/prometheus-operator-values.yaml (hunk 10 above)
Comment (h314to) [reply docs/configuration/git-providers/monitoring-conf/prometheus-operator-values.yaml]: Resolution is made in the same namespace where grafana is running, so no.

File: docs/configuration/monitoring.md (hunk 8 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: That's strage. I'm based on `ca2d5bb`, "Sandbox/slack artifact url", and there was only commit after that, which didn't change remark. Codacy also didn't pickup anything. But I'll check.

File: docs/configuration/monitoring.md (hunk 7 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: Yes, I'll make it more legible. Note that Grafana and prometheus are installed by the [same chart](https://github.com/helm/charts/tree/master/stable/prometheus-operator).

Given that the documentation for monitoring is currently not vast it is probably not too hard to follow. Having the docs for crow and prometheus+loki+grafana in the same place has the virtue of making it easier for a user to get info about the monitoring options available.

We could make it clearer by changing the text at the top of the document, and possibly inverting the order, so that the simpler and built-in solution (crow) is documented first. I'll tweak it a bit.

File: docs/configuration/monitoring.md [hunk 11]
```diff
@@ -1,43 +1,142 @@
...
-  --version 6.9.3 \
-  --namespace monitoring \
-  --set prometheusOperator.createCustomResource=false \
-  --set grafana.service.type=NodePort\
-  --set grafana.adminPassword="CHANGE_HERE"
+```bash
+helm repo add loki https://grafana.github.io/loki/charts
 
-kubectl get svc monitoring-grafana -n monitoring
+kubectl create namespace monitoring
```
Comment (paulopontesm) [inline docs/configuration/monitoring.md]:
```suggestion
```

Comment (paulopontesm) [reply docs/configuration/monitoring.md]: I think this was meant to be deleted 🤔

Comment (h314to) [reply docs/configuration/monitoring.md]: Done. Thanks. Forgot that one :)

File: docs/configuration/monitoring.md [hunk 12]
```diff
@@ -1,43 +1,140 @@
...
-In older versions, Codacy was monitored using an application we build
-in-house called "Crow". Going forward, we are deprecating Crow and
-setting up "Prometheus style" metrics in all our components that
-can be accessed using the Prometheus + Grafana ecosystem.
+Currently two monitoring solutions are supported:
 
-Initially Crow will still be bundled in Codacy helm chart, until
-we have all functionality available on Grafana.
+* [**Crow**](##Setting-up-monitoring-using-Crow) - A simple, lightweight, and builtin monitoring solution.
+* [**Prometheus+Grafana+Loki**](##Setting-up-monitoring-using-Grafana,-Prometheus,-and-Loki) - A more comprehensive third-party monitoring solution, for more advanced usage.
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: Fix links and minor edits:

```suggestion
* [**Crow**](#setting-up-monitoring-using-crow) - A simple, lightweight, and built-in monitoring solution.
* [**Prometheus + Grafana + Loki**](#setting-up-monitoring-using-grafana-prometheus-and-loki) - A comprehensive third-party monitoring solution, recommended for more advanced usage.
```

File: docs/configuration/monitoring.md [hunk 13]
```diff
@@ -1,43 +1,140 @@
...
 
-## Setting up monitoring using Grafana & Prometheus
+## Setting up monitoring using Crow
+
+Crow is a visualization tool that displays information about the projects and jobs
+that are pending for analysis, as well as the ones running in the Codacy platform.
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: Just to clarify, I interpret this as "Crow displays information about the **projects pending analysis** and the **jobs currently running** on Codacy". Is this correct?

File: docs/configuration/monitoring.md [hunk 14]
```diff
@@ -1,43 +1,140 @@
...
+For example:
+
+```yaml
+global:
+  codacy:
+    crow:
+      url: "http://codacy.example.com/monitoring"
+```
+
+Please see the [README.md](https://github.com/codacy/chart/blob/master/README.md) for more information about these values.
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: I think we can drop this detail, as we are not mentioning the values reference in the `README.md` for the other .yaml values files.

Advanced users who may want to change the values besides the ones we identify with placeholders will probably know where to find the information, no?

File: docs/configuration/monitoring.md [hunk 15]
```diff
@@ -1,43 +1,140 @@
...
+
+### Configuring your credentials
+
+You **should** provide a password for the `crow` installation either through the `values.yaml` file or through a `--set` parameter during the `helm` installation process. This parameter can be configured as follows:
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: Defining the custom password is not mandatory (since there is a default one) but highly recommended, is this right? Also, to improve the documentation readability we can simply use the capitalized proper nouns of the tools throughout this page, when appropriate of course:

```suggestion
We highly recommend that you define a custom password for the Crow installation. You can do this either through the `values.yaml` file or through a `--set` parameter during the Helm installation process. This parameter can be configured as follows:
```

File: docs/configuration/monitoring.md [hunk 16]
```diff
@@ -1,43 +1,140 @@
...
+password: C0dacy123
+```
+
+## Setting up monitoring using Grafana, Prometheus, and Loki
+
+[Prometheus](https://prometheus.io) is an open-source systems monitoring and alerting
+toolkit. Logs can be collected using [Loki](https://grafana.com/oss/loki/), which is a
+horizontally-scalable, highly-available, multi-tenant log aggregation system
+It's data can be visualized with [Grafana](https://grafana.com), a widely used
+open source analytics and monitoring solution.
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]:
```suggestion
Its data can be visualized with [Grafana](https://grafana.com), a widely used
open-source analytics and monitoring solution.
```

File: docs/configuration/monitoring.md [hunk 17]
```diff
@@ -1,43 +1,140 @@
...
+
+The following guide covers the basic installation of the components. This monitoring
+stack is considerably more resource demanding than Crow, and is recommended only for
+more advanced usage. Furthermore, its installation, configuration, and management
+requires a deeper knowledge of Kubernetes as each component must be carefully tweaked
+to match your specific use case, using as starting point the values files herein provided.
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]:
```suggestion
The following guide covers the basic installation of the components in this monitoring stack.
This solution is considerably more resource demanding than Crow, and is recommended only for
more advanced usage. Furthermore, its installation, configuration, and management
requires a deeper knowledge of Kubernetes as each component must be carefully tweaked
to match your specific use case, using as starting point the .yaml values files provided by us.
```

File: docs/configuration/monitoring.md [hunk 18]
```diff
@@ -1,43 +1,140 @@
...
+
+### Installing CRDs
 
 The simplest way to setup prometheus in your cluster is by using the
 [Prometheus-Operator](https://github.com/helm/charts/tree/master/stable/prometheus-operator)
-bundle.
+bundle. Start by adding the custom resources required for installing this bundle in your cluster.
+
+NOTE: if installing on MicroK8s use `microk8s.kubectl` instead of `kubectl`
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]:
```suggestion
**NOTE:** if installing on MicroK8s use `microk8s.kubectl` instead of `kubectl`
```

Comment (pauloribeiro-codacy) [reply docs/configuration/monitoring.md]: @paulopontesm The note instructing users to run `microk8s.kubectl` is getting spread out throughout the documentation. Maybe it's better to recommend using the alias in the MicroK8s-specific quick guide? (This would be something for a different pull request, though.)

File: docs/configuration/monitoring.md [hunk 19]
```diff
@@ -1,43 +1,140 @@
...
+It's data can be visualized with [Grafana](https://grafana.com), a widely used
+open source analytics and monitoring solution.
+
+The following guide covers the basic installation of the components. This monitoring
+stack is considerably more resource demanding than Crow, and is recommended only for
+more advanced usage. Furthermore, its installation, configuration, and management
+requires a deeper knowledge of Kubernetes as each component must be carefully tweaked
+to match your specific use case, using as starting point the values files herein provided.
+
+### Installing CRDs
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: What are CRDs? :nerd_face: I genuinely don't know, and there's a chance that customers may not know as well.

File: docs/configuration/monitoring.md [hunk 20]
```diff
@@ -1,43 +1,140 @@
...
+```
+
+### Install Loki
 
-sleep 5 # wait for crds to be created
+Obtain the configuration file for Loki, [values-loki.yaml](https://github.com/codacy/chart/blob/master/codacy/values-loki.yaml),
+and install it as described below. You may need to edit the storage class for Loki
+persistence in this configuration file. While the default setting should suit most use
+cases, you may need to adjust it to your specific Kubernetes installation.
+For instance, for MicroK8s use `storageClassName: microk8s-hostpath`.
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]:
```suggestion
Obtain the configuration file for Loki, [values-loki.yaml](https://github.com/codacy/chart/blob/master/codacy/values-loki.yaml),
and install it as described below. While the default storage class setting
for Loki persistence should suit most use cases, you may need to adjust
it to your specific Kubernetes installation. For instance, for MicroK8s use
`storageClassName: microk8s-hostpath`.
```

File: docs/configuration/monitoring.md [hunk 21]
```diff
@@ -52,51 +149,10 @@ engine:
...
-
-## Default credentials
-
-If you have not configured your `crow` credentials as described [above](###Configuring-your-credentials), you can login with the following default credentials:
-
-```yaml
-username: codacy
-password: C0dacy123
+```bash
+helm upgrade (...options used to install codacy...) --values values-monitoring-values.yaml --recreate-pods
```
Comment (pauloribeiro-codacy) [inline docs/configuration/monitoring.md]: We have a similar `helm upgrade` command above for the Crow credentials but we aren't explaining how the user should complete the command like we do here. Should we choose to provide more detailed instructions like here, or less like above? :thinking:

**Comment by paulopontesm** [comment]
These changes are being brought in this PR: https://github.com/codacy/chart/pull/332/files

File: docs/configuration/monitoring.md (hunk 13 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: I barely touched the Crow documentation. I just moved it to the top of the document. It was changed just a week ago so I assumed it was thoroughly reviewed then. These sentences seem equivalent, but I can change it nonetheless.

File: docs/configuration/monitoring.md (hunk 14 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: As above. Crow isn't really for advanced users, but I also see no need in this line. I'll remove it.

File: docs/configuration/monitoring.md (hunk 15 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: "Should" is generally used to convey a recommendation or an expected behaviour, and doesn't imply something is mandatory.

File: docs/configuration/monitoring.md (hunk 16 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: Good catch for the typo in the possessive. Thanks. Regarding `open source` I note that the authorities on the subject, e.g. the  [Open Source Initiative](https://opensource.org), do not hyphenate.

File: docs/configuration/monitoring.md (hunk 17 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: Herein is more informative because, while it implies that the files are provided by us, it also states that they are in this repo and not elsewhere. The rest is somewhat equivalent, but I don't mind changing it.

File: docs/configuration/monitoring.md (hunk 19 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: They are [Custom Resource Definitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/), i.e. extensions of the Kubernetes API. I surely hope that people are not installing CRDs without knowing what they are, and as we stated above, this stack is for more advanced users. But I don't mind adding a link to the documentation.

File: docs/configuration/monitoring.md (hunk 21 above)
Comment (h314to) [reply docs/configuration/monitoring.md]: I'd rather add more info to the command above. `(...)` alone isn't often used to document command line tools, and its meaning might be obscure.
